docs/index.rst

```diff
@@ -0,0 +1,212 @@
+no type bindings, no XML configurations.
+
+.. __: https://en.wikipedia.org/wiki/Dependency_injection
+
+
+Why?
+----
+
+Dependency Injection (DI) design pattern is intended to decouple various parts
+of an application from each other. So a class can be independent of how the
+objects it requires are created, and hence the way we create them may be
+different for production and tests.
+
+One of most easiest examples is to say that DI is essentially about writing
+
+.. code:: python
+
+    def do_something(my_service):
+        return my_service.get_val() + 42
+
+    my_service = MyService(foo, bar)
+    do_something(my_service)
+
+instead of
+
+.. code:: python
+
+    def do_something():
+        my_service = MyService(foo, bar)
+        return my_service.get_val() + 42
+
+    do_something()
+
+because the latter considered non-configurable and is harder to test.
+
+In Python, however, dependency injection is not a big deal due to its dynamic
+nature and duck typing: anything could be defined anytime and passed anywhere.
+Due to that reason (and maybe some others) DI frameworks aren't popular among
+Python community, though they may be handy in some cases.
+
+One of such cases is code decoupling when we want to create objects in one
+places and use them in others, preserving clean interface and avoiding global
+variables. Having all these considerations in mind, Picobox was born.
+
+
+Quickstart
+----------
+
+Picobox provides ``Box`` class that acts as a container for objects you want
+to deal with. You can put, you can get, you can pass them around.
+
+.. code:: python
+
+    import picobox
+
+    box = picobox.Box()
+    box.put('foo', 42)
+
+    @box.pass_('foo')
+    def spam(foo):
+        return foo
+
+    @box.pass_('foo', as_='bar')
+    def eggs(bar):
+        return bar
+
+    print(box.get('foo'))   # 42
+
+    print(spam())           # 42
+    print(eggs())           # 42
+
+One of the key principles is to `do not break` existing code. That's why
+Picobox does not change function signature and injects dependencies as if they
+are defaults.
+
+.. code:: python
+
+    print(spam())           # 42
+    print(spam(13))         # 13
+    print(spam(foo=99))     # 99
+
+Another key principle is that ``pass_()`` resolves dependencies lazily which
+means you can inject them everywhere you need and define them much later. The
+only rule is to define them before calling the function.
+
+.. code:: python
+
+    import picobox
+
+    box = picobox.Box()
+
+    @box.pass_('foo')
+    def spam(foo):
+        return foo
+
+    print(spam(13))         # 13
+    print(spam())           # KeyError: 'foo'
+
+    box.put('foo', 42)
+
+    print(spam())           # 42
+
+The value to inject is not necessarily an object. You can pass a factory
+function which will be used to produce a dependency. A factory function has
+no arguments, and it's assumed it has all the context it needs to work.
```

Comment:
Better to say in passive form: "and is assumed to have all the context it needs to work".